plugins/Core-Extensions/Recommended/edu.toronto.cs.se.modelepedia.operator/src/edu/toronto/cs/se/modelepedia/operator/slice/Slice.java

```diff
@@ -74,20 +76,181 @@ public Input(Map<String, Model> inputsByName) {
+	// between elements in the unchecked list and the criterion list.
+	public List<EObject> matchCriterionToModelElements(
+			List<EObject> unchecked, List<EObject> criterionList) {
+		List<EObject> unprocessed = new ArrayList<EObject>();
+
+		Iterator<EObject> uncheckedIter = unchecked.iterator();
+		while (uncheckedIter.hasNext()) {
+			EObject elem = uncheckedIter.next();
+			for (EObject critElem : criterionList) {
+				if (isModelElemEqual(elem, critElem)) {
+					unprocessed.add(elem);
+					System.out.println("Criterion Added: " + elem);
+					uncheckedIter.remove();
+				}
+			}
+		}
+		
+		return unprocessed;
+	}
+	
+	// Checks whether the two input model elements are equivalent. 
+	// This is necessary as even for the same model elements, different 
+	// calls to getEMFInstanceObject() will return different EMF instances.
+	public boolean isModelElemEqual(EObject obj1, EObject obj2) {
+		return true;
+	}
+	
+	// Checks whether the elements in the two input lists are equivalent.
+	// Note: It is assumed that there must be a one-to-one mapping between
+	// each element in the two input lists.
+	public boolean isElemListEqual(List<? extends EObject> list1, 
+			List<? extends EObject> list2) {
+		boolean equalFlag = true; 
+
+		for (EObject obj1 : list1) {
+			boolean curEqualFlag = false;
+			Iterator<? extends EObject> iter2 = list2.iterator();
+			
+			while (iter2.hasNext()) {
+				if (isModelElemEqual(obj1, iter2.next())) {
+					iter2.remove();
+					curEqualFlag = true;
+					break;						
+				}
+			}
+			
+			equalFlag = equalFlag && curEqualFlag;
+		}
+		
+		
+		if (!list2.isEmpty()) {
+			equalFlag = false;
+		}
+		
+		return equalFlag;
+		
+				
+	}
+
+	// Returns the complete list of model elements that may be impacted
+	// by the model elements included in the original slicing criterion.
+	// By default, all model elements are assumed to be impacted by the slice.
+	public List<EObject> getImpactedElements(
+			List<EObject> unprocessed, List<EObject> unchecked) {
+		
+		List<EObject> changed = new ArrayList<EObject>();
+		Iterator<EObject> unprocessedIter = unprocessed.iterator();
+		
+		while (unprocessedIter.hasNext()) {
+			System.out.println("Elements to be processed: " + 
+					+ unprocessed.size());
+			EObject obj = unprocessedIter.next();
+
+			if (changed.contains(obj)){
+				// This element has already been checked for dependencies.
+				unprocessed.remove(obj);
+
+			} else {
+				unprocessed.remove(obj);
+				changed.add(obj);
+				System.out.println("Current element: " + obj);
+				
+				// Get all model elements affected by the impacted model element.
+				EObject uncheckedObj;
+				Iterator<EObject> uncheckedIter = unchecked.iterator();
+				while (uncheckedIter.hasNext()) {
+					uncheckedObj = uncheckedIter.next();
+					if (isModelElemImpactedBy(uncheckedObj, obj)) {
+						unprocessed.add(uncheckedObj);
+						uncheckedIter.remove();
+					}
+				}
+			}
+			
+			unprocessedIter = unprocessed.iterator();
```

Comment:
This smells like it has the potential to go into an infinite loop, or at least it's a very peculiar way of doing recursion.